FontCascadeCache: ignore FontSelector ID and Generation for "Simple" cases

vitorroriz opened the pull request:
#### 492c00cb99676770931d9a9ea091e14488707306
<pre>
FontCascadeCache: ignore FontSelector ID and Generation for &quot;Simple&quot; cases
<a href="https://bugs.webkit.org/show_bug.cgi?id=288756">https://bugs.webkit.org/show_bug.cgi?id=288756</a>
<a href="https://rdar.apple.com/145786886">rdar://145786886</a>

Reviewed by Matthieu Dubet and Geoffrey Garen.

We currently have a FontCascadeCache that stores FontCascadeFont objects that can be
retrieved when we resolve style. The key for this is composed by:

1. FontDescriptionKey -&gt; based on the FontCascadeDescription
2. Font family names -&gt; based on FontCascadeDescription
3. CSSFontSelector unique ID -&gt; unique ID is “statically incremented” every time a selector is constructed
4. CSSFontSelector generation -&gt; generation changes every time we process a rule, or start building the selector

Generation and ID change constantly, even for equivalent CSSFontSelectors.
It is not cheap to tell two font selectors are equivalent for all cases as we would have to consider all the possible
descriptors for all rules (@font-faces, palettes, font feature values).

However, if we can tell that a selector is &quot;simple&quot; we should be able to ignore all these factors.

The idea is to classify a CSSFontSelector as “simple”. Simple meaning that there are no author rules attached to it.
When this happens we compose the key for fetching FontCascadeFont with ID and generation 0, which is the same for when
we don’t have a CSSFontSelector. We still use all font-family names and FontDescription information, of course.

* LayoutTests/TestExpectations:
* LayoutTests/imported/blink/fast/text/update-sans-serif-and-recalc-style.html:
* LayoutTests/platform/mac-wk1/TestExpectations:
* LayoutTests/platform/mac/fast/css/font-face-default-font-expected.png:
* LayoutTests/platform/mac/svg/W3C-I18N/tspan-dirRTL-ubEmbed-in-default-context-expected.png:
* LayoutTests/platform/mac/svg/W3C-I18N/tspan-dirRTL-ubEmbed-in-ltr-context-expected.png:
* Source/WebCore/css/CSSCounterStyleRegistry.h:
(WebCore::CSSCounterStyleRegistry::hasAuthorCounterStyles const):
* Source/WebCore/css/CSSFontSelector.cpp:
(WebCore::CSSFontSelector::isSimpleFontSelectorForDescription const):
* Source/WebCore/css/CSSFontSelector.h:
(isType):
* Source/WebCore/platform/graphics/FontCascadeCache.cpp:
(WebCore::makeFontCascadeCacheKey):
* Source/WebCore/platform/graphics/FontCascadeCache.h:
* Source/WebCore/platform/graphics/FontSelector.h:
(WebCore::FontSelector::isCSSFontSelector const):

Canonical link: <a href="https://commits.webkit.org/291808@main">https://commits.webkit.org/291808@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/daf72979c936376e2e42147e0bd65ae53600b601

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/94090 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/13677 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/3423 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/99101 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/44617 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/96140 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/13977 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/22107 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/71787 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/29127 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/97092 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/132/builds/10371 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/84963 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/52128 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/10055 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/2637 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/43933 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/80301 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/2724 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/101140 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/21142 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/15396 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/80793 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/21394 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/80963 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/80168 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/19973 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/24710 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/137/builds/2074 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/14304 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/21126 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/26305 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/125/builds/20813 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/24273 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/22554 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

rreno commented on LayoutTests/TestExpectations:
```diff
@@ -7749,3 +7749,6 @@ webkit.org/b/286883 imported/w3c/web-platform-tests/css/CSS2/floats/remove-block
 [ Debug ] imported/w3c/web-platform-tests/css/CSS2/floats/line-pushed-by-floats-crash.html [ Skip ]
+
+# This test is flaky. Resizing the window exposures the failure
+webkit.org/b/288689 svg/text/bidi-text-query.svg [ Failure Pass ImageOnlyFailure ]
```
Is there any value to running this test at all? If not, can we just skip it until the bug is fixed?

brentfulgham commented on Source/WebCore/css/CSSFontSelector.cpp:
```diff
@@ -445,4 +446,19 @@ RefPtr<Font> CSSFontSelector::fallbackFontAt(const FontDescription& fontDescript
+            return false;
+    }
+
+    return !m_cssFontFaceSet->faceCount() && m_featureValues.isEmpty() && m_paletteMap.isEmpty();
```
I wonder if putting the `faceCount()` call that involves a pointer dereference to the end of the conditionals would squeeze any more perf out since the local container empty checks would fail the && operation immediately

vitorroriz replied:
This is a good point. My reasoning here (based only on intuition and no data) was that `@font-face` is more common than `m_featureValues` and `m_paletteMap`, which are usually empty. So if we see an `@font-face` rule we would already return false.

I'd imagine for SP3 all of these are usually empty but that's something we can easily measure on an A/B.

mdubet replied:
We could early return separately for fontFace, featureValues and paletteMap to make it slightly more readable and certainly slightly faster.

The counter style fix should be the last one also (certainly the most costly).

brentfulgham commented on LayoutTests/imported/w3c/web-platform-tests/css/css-values/lh-unit-005-expected.txt:
```diff
@@ -1,3 +1,3 @@
 
-FAIL lh relative to line-height inherited from parent assert_equals: expected "100px" but got "50px"
+PASS lh relative to line-height inherited from parent
```
Nice! Perf progression AND correctness progression.

geoffreygaren commented on Source/WebCore/platform/graphics/FontCascadeCache.cpp:
```diff
@@ -95,20 +95,26 @@ void FontCascadeCache::pruneSystemFallbackFonts()
+    if (!addResult.isNewEntry) {
+        auto fonts = addResult.iterator->value->fonts;
+        if (!key.hasComplexFontSelector)
+            fonts->updateFontSelector(WTFMove(fontSelector));
```
I think I see a bug here: As you explained to me over WebEx, you're updating the `FontSelector` here because it points to a `ScriptExecutionContext` (`Document`) that can differ from frame to frame.

To the extent that two frames might have two different documents, but transitively point to the same `FontSelector`, we can't solve that conflict by letting the last cache lookup win, since the previous frame might continue to use its `FontCascadeFonts` after the next frame updates the `FontSelector`'s document.

Instead, I think we need to change `FontSelector` to remove its stored `m_context` and require `ScriptExecutionContext` or `Settings` as a function argument anywhere it will need them. Then `FontSelector` (and therefore `FontCascadeFonts`) can be context-free, and shared between documents.

geoffreygaren reviewed (approved):
This is great!

I'm going to say r+ now because this patch doesn't need another round of review later, but please resolve these two issues before landing:
* Please remove the line of code that updates m_context when retrieving items from the cache
* Please land the separate patch we talked about to remove m_context from CSSFontSelector (explained more below)

mdubet commented on Source/WebCore/css/CSSFontSelector.cpp:
```diff
@@ -445,4 +446,19 @@ RefPtr<Font> CSSFontSelector::fallbackFontAt(const FontDescription& fontDescript
+    if (m_stagingArea.size())
+        return false;
+
+    // FIXME: remove this when we fix counter style rules mutation.
```
Do we have a bugzilla for this ?

mdubet commented on Source/WebCore/css/CSSFontSelector.cpp:
```diff
@@ -445,4 +446,19 @@ RefPtr<Font> CSSFontSelector::fallbackFontAt(const FontDescription& fontDescript
+bool CSSFontSelector::isSimpleFontSelectorForDescription() const
+{
+    // font face rules still pending
+    if (m_stagingArea.size())
```
We generally use `!...isEmpty()`

mdubet reviewed (approved):
LGTM. I'm a bit skeptical about the `updateFontSelector()` when we retrieve a cache entry ? What is the reason ?

geoffreygaren reviewed (changes requested):
Upon reflection, the original plan didn't work. The change to extract `ScriptExecutionContext` from `FontSelector` has proved challenging, and `CSSFontSelector` also inherits `ActiveDOMObject`, which is a per-context concept. `CSSFontSelector` also has a `FontSelectorClient` that needs to learn about invalidation, and it's not clear what happens if a body of code uses a cached `CSSFontSelector` for which it has not registered as a client. 

We need to identify a strategy for an object or a cache that separates itself from per-context state.

geoffreygaren commented on Source/WebCore/platform/graphics/FontCascadeCache.cpp:
```diff
@@ -95,20 +97,35 @@ void FontCascadeCache::pruneSystemFallbackFonts()
+        auto fonts = addResult.iterator->value->fonts;
+        if (!key.hasComplexFontSelector) {
+            if (RefPtr cachedSelector = dynamicDowncast<CSSFontSelector>(fonts->fontSelector()); cachedSelector && cachedSelector->isContextStopped()) {
+                // We just produce produce a hit and replace the cached FontCascadeFont's FontSelector if its context is "stopped"
```
What happens if a context is stopped and then restarted?

vitorroriz replied:
I believe `m_activeDOMObjectsAreStopped` is never set to false. It is initialized to false and set to true during teardown.

It seems we have a separated flag for the suspended state `m_activeDOMObjectsAreSuspended`, which can be resumed, setting `m_activeDOMObjectsAreSuspended` to true (`resumeActiveDOMObjects`)

geoffreygaren reviewed (approved):
Yes, I think that change addresses my earlier comment. Maybe check in with other reviewers on their style comments before landing.

vitorroriz committed:
FontCascadeCache: ignore FontSelector ID and Generation for "Simple" cases
https://bugs.webkit.org/show_bug.cgi?id=288756
rdar://145786886

Reviewed by Matthieu Dubet and Geoffrey Garen.

We currently have a FontCascadeCache that stores FontCascadeFont objects that can be
retrieved when we resolve style. The key for this is composed by:

1. FontDescriptionKey -> based on the FontCascadeDescription
2. Font family names -> based on FontCascadeDescription
3. CSSFontSelector unique ID -> unique ID is “statically incremented” every time a selector is constructed
4. CSSFontSelector generation -> generation changes every time we process a rule, or start building the selector

Generation and ID change constantly, even for equivalent CSSFontSelectors.
It is not cheap to tell two font selectors are equivalent for all cases as we would have to consider all the possible
descriptors for all rules (@font-faces, palettes, font feature values).

However, if we can tell that a selector is "simple" we should be able to ignore all these factors.

The idea is to classify a CSSFontSelector as “simple”. Simple meaning that there are no author rules attached to it.
When this happens we compose the key for fetching FontCascadeFont with ID and generation 0, which is the same for when
we don’t have a CSSFontSelector. We still use all font-family names and FontDescription information, of course.

* LayoutTests/TestExpectations:
* LayoutTests/imported/blink/fast/text/update-sans-serif-and-recalc-style.html:
* LayoutTests/platform/mac-wk1/TestExpectations:
* LayoutTests/platform/mac/fast/css/font-face-default-font-expected.png:
* LayoutTests/platform/mac/svg/W3C-I18N/tspan-dirRTL-ubEmbed-in-default-context-expected.png:
* LayoutTests/platform/mac/svg/W3C-I18N/tspan-dirRTL-ubEmbed-in-ltr-context-expected.png:
* Source/WebCore/css/CSSCounterStyleRegistry.h:
(WebCore::CSSCounterStyleRegistry::hasAuthorCounterStyles const):
* Source/WebCore/css/CSSFontSelector.cpp:
(WebCore::CSSFontSelector::isSimpleFontSelectorForDescription const):
* Source/WebCore/css/CSSFontSelector.h:
(isType):
* Source/WebCore/platform/graphics/FontCascadeCache.cpp:
(WebCore::makeFontCascadeCacheKey):
* Source/WebCore/platform/graphics/FontCascadeCache.h:
* Source/WebCore/platform/graphics/FontSelector.h:
(WebCore::FontSelector::isCSSFontSelector const):

Canonical link: https://commits.webkit.org/291808@main
